sa/metrics.go

```diff
@@ -1,125 +1,97 @@
-
-	waitDuration := prometheus.NewCounter(prometheus.CounterOpts{
-		Name: "db_wait_duration_seconds",
-		Help: "The total time blocked waiting for a new connection.",
-	})
-	scope.MustRegister(waitDuration)
-
-	maxIdleClosed := prometheus.NewCounter(prometheus.CounterOpts{
-		Name: "db_max_idle_closed",
-		Help: "Total number of connections closed due to SetMaxIdleConns.",
-	})
-	scope.MustRegister(maxIdleClosed)
-
-	maxLifetimeClosed := prometheus.NewCounter(prometheus.CounterOpts{
-		Name: "db_max_lifetime_closed",
-		Help: "Total number of connections closed due to SetConnMaxLifetime.",
-	})
-	scope.MustRegister(maxLifetimeClosed)
+// Describe is implemented with DescribeByCollect. That's possible because the
+// Collect method will always return the same metrics with the same descriptors.
+func (dbc dbMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	prometheus.DescribeByCollect(dbc, ch)
+}
 
-	// Construct a dbMetrics instance with all of the registered metrics and the
-	// gorp DBMap
-	return &dbMetrics{
-		dbMap:              dbMap,
-		maxOpenConnections: maxOpenConns,
-		openConnections:    openConns,
-		inUse:              inUse,
-		idle:               idle,
-		waitCount:          waitCount,
-		waitDuration:       waitDuration,
-		maxIdleClosed:      maxIdleClosed,
-		maxLifetimeClosed:  maxLifetimeClosed,
+// Collect first triggers the dbMaps's sql.Db's Stats function. Then it
+// creates constant metrics for each DBStats value on the fly based on the
+// returned data.
+//
+// Note that Collect could be called concurrently, so we depend on
+// Stats() to be concurrency-safe.
+func (dbc dbMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	writeStat := func(stat *prometheus.Desc, typ prometheus.ValueType, val float64) {
+		ch <- prometheus.MustNewConstMetric(stat, typ, val)
+	}
+	writeCounter := func(stat *prometheus.Desc, val float64) {
+		writeStat(stat, prometheus.CounterValue, val)
+	}
+	writeGauge := func(stat *prometheus.Desc, val float64) {
+		writeStat(stat, prometheus.GaugeValue, val)
 	}
-}
 
-// updateFrom updates the dbMetrics prometheus stats based on the provided
-// sql.DBStats object.
-func (dbm *dbMetrics) updateFrom(dbStats sql.DBStats) {
-	dbm.maxOpenConnections.Set(float64(dbStats.MaxOpenConnections))
-	dbm.openConnections.Set(float64(dbStats.OpenConnections))
-	dbm.inUse.Set(float64(dbStats.InUse))
-	dbm.idle.Set(float64(dbStats.InUse))
-	dbm.waitCount.Set(float64(dbStats.WaitCount))
-	dbm.waitDuration.Set(float64(dbStats.WaitDuration.Seconds()))
-	dbm.maxIdleClosed.Set(float64(dbStats.MaxIdleClosed))
-	dbm.maxLifetimeClosed.Set(float64(dbStats.MaxLifetimeClosed))
+	// Translate the DBMap's db.DBStats counter values into Prometheus metrics.
+	dbMapStats := dbc.dbMap.Db.Stats()
+	writeGauge(maxOpenConns, float64(dbMapStats.MaxOpenConnections))
+	writeGauge(openConns, float64(dbMapStats.OpenConnections))
+	writeGauge(inUse, float64(dbMapStats.InUse))
+	writeGauge(idle, float64(dbMapStats.InUse))
```

Comment:
```suggestion
	writeGauge(idle, float64(dbMapStats.Ide))
```

Looks like we had a copy-paste bug in the old code; might as well fix as part of this PR.

Reply:
Good catch, thanks! Fixed in 9d4709e